### DISCUSSION TITLE
One year of NNUE....

### DISCUSSION THREAD

**Discussion by vondele** [official-stockfish/Stockfish#3628] [Show and tell]
We have merged NNUE about a year ago, so time for some retrospective on the incredible progress.

I'll kick it off by a graph showing how much search speed improved since the first merged version (already optimized by itself):

![NNUE_Mnps](https://user-images.githubusercontent.com/4202567/126898745-599895e6-08b3-4a3f-8e73-bb5e354ede19.png)

**Comment by Sopel97** [comment]
The commits with largest speedups:
https://github.com/official-stockfish/Stockfish/commit/ea6220f3813e5b76b444a02905eaf2c556bdb368
https://github.com/official-stockfish/Stockfish/commit/8b8a510fd6a1a17b39b2d4b166f60ac7be0dab23
https://github.com/official-stockfish/Stockfish/commit/c065abdcafe0486cb5cfa7de12a4ac6a905a54c5
https://github.com/official-stockfish/Stockfish/commit/2046d5da30b2cd505b69bddb40062b0d37b43bc7
https://github.com/official-stockfish/Stockfish/commit/d862ba40692797031ec5b0d95e46bcfc5a80f06c
https://github.com/official-stockfish/Stockfish/commit/23c385ec36f9d5a9514ec5b0811ec99d08b45e90

**Comment by vondele** [comment]
We have also had 28 different default nets in about a year time. So on average 2 weeks between net releases:

I have conducted matches for all of these 28 version against SF_classical (9587eeeb5ed29f834d4f956b92e0e732877c47a7) the latest commit before the merge of NNUE, using STC TC (10+0.1) and the noob_3moves.epd book.

1 84f3e86 2020-08-06 [nn-97f742aaefcd.nnue](https://tests.stockfishchess.org/api/nn/nn-97f742aaefcd.nnue) 81.5 +- 3.9 : 25000 (+9872,=10967,-4161), 61.4 %
2 857e045 2020-08-07 [nn-9931db908a9b.nnue](https://tests.stockfishchess.org/api/nn/nn-9931db908a9b.nnue) 122.3 +- 3.5 : 25000 (+11324,=10735,-2941), 66.8 %
3 bcdf41d 2020-08-10 [nn-112bb1c8cdb5.nnue](https://tests.stockfishchess.org/api/nn/nn-112bb1c8cdb5.nnue) 127.2 +- 3.5 : 25000 (+11437,=10823,-2740), 67.4 %
4 67e4841 2020-08-13 [nn-82215d0fd0df.nnue](https://tests.stockfishchess.org/api/nn/nn-82215d0fd0df.nnue) 172.8 +- 3.4 : 25000 (+13215,=9985,-1800), 72.8 %
5 9cc482c 2020-09-04 [nn-308d71810dff.nnue](https://tests.stockfishchess.org/api/nn/nn-308d71810dff.nnue) 177.4 +- 3.4 : 25000 (+13495,=9683,-1822), 73.3 %
6 7135678 2020-09-15 [nn-03744f8d56d8.nnue](https://tests.stockfishchess.org/api/nn/nn-03744f8d56d8.nnue) 207.4 +- 3.4 : 25000 (+14580,=9115,-1305), 76.5 %
7 36c2886 2020-09-28 [nn-04a843f8932e.nnue](https://tests.stockfishchess.org/api/nn/nn-04a843f8932e.nnue) 217.0 +- 3.6 : 25000 (+15113,=8536,-1351), 77.5 %
8 5efbaab 2020-09-28 [nn-baeb9ef2d183.nnue](https://tests.stockfishchess.org/api/nn/nn-baeb9ef2d183.nnue) 221.6 +- 3.6 : 25000 (+15208,=8572,-1220), 78.0 %
9 ba73f8c 2020-10-14 [nn-04cf2b4ed1da.nnue](https://tests.stockfishchess.org/api/nn/nn-04cf2b4ed1da.nnue) 235.7 +- 3.7 : 25000 (+15771,=8122,-1107), 79.3 %
10 281d520 2020-10-18 [nn-eba324f53044.nnue](https://tests.stockfishchess.org/api/nn/nn-eba324f53044.nnue) 235.1 +- 3.8 : 25000 (+15793,=8049,-1158), 79.3 %
11 6328135 2020-10-28 [nn-2eb2e0707c2b.nnue](https://tests.stockfishchess.org/api/nn/nn-2eb2e0707c2b.nnue) 226.3 +- 3.6 : 25000 (+15435,=8343,-1222), 78.4 %
12 dfc7f88 2020-11-01 [nn-cb26f10b1fd9.nnue](https://tests.stockfishchess.org/api/nn/nn-cb26f10b1fd9.nnue) 236.0 +- 3.9 : 25000 (+15814,=8049,-1137), 79.4 %
13 32edb1d 2020-11-08 [nn-c3ca321c51c9.nnue](https://tests.stockfishchess.org/api/nn/nn-c3ca321c51c9.nnue) 241.9 +- 3.8 : 25000 (+16025,=7901,-1074), 79.9 %
14 7364006 2020-11-29 [nn-62ef826d1a6d.nnue](https://tests.stockfishchess.org/api/nn/nn-62ef826d1a6d.nnue) 246.0 +- 3.9 : 25000 (+16140,=7859,-1001), 80.3 %
15 e8d64af 2021-05-18 [nn-8a08400ed089.nnue](https://tests.stockfishchess.org/api/nn/nn-8a08400ed089.nnue) 303.9 +- 4.4 : 25000 (+18275,=5947,-778), 85.0 %
16 fb2d175 2021-05-22 [nn-7756374aaed3.nnue](https://tests.stockfishchess.org/api/nn/nn-7756374aaed3.nnue) 309.5 +- 4.4 : 25000 (+18416,=5862,-722), 85.4 %
17 d53071e 2021-06-03 [nn-7e66505906a6.nnue](https://tests.stockfishchess.org/api/nn/nn-7e66505906a6.nnue) 307.7 +- 4.4 : 25000 (+18363,=5898,-739), 85.2 %
18 f8c779d 2021-06-14 [nn-8e47cf062333.nnue](https://tests.stockfishchess.org/api/nn/nn-8e47cf062333.nnue) 313.1 +- 4.4 : 25000 (+18559,=5701,-740), 85.6 %
19 8ec9e10 2021-06-15 [nn-33c9d39e5eb6.nnue](https://tests.stockfishchess.org/api/nn/nn-33c9d39e5eb6.nnue) 331.9 +- 4.6 : 25000 (+19044,=5370,-586), 86.9 %
20 14b673d 2021-06-18 [nn-3b20abec10c1.nnue](https://tests.stockfishchess.org/api/nn/nn-3b20abec10c1.nnue) 330.8 +- 4.5 : 25000 (+19021,=5371,-608), 86.8 %
21 86afb6a 2021-06-18 [nn-aa9d7eeb397e.nnue](https://tests.stockfishchess.org/api/nn/nn-aa9d7eeb397e.nnue) 324.3 +- 4.5 : 25000 (+18836,=5532,-632), 86.4 %
22 adfb23c 2021-06-18 [nn-50144f835024.nnue](https://tests.stockfishchess.org/api/nn/nn-50144f835024.nnue) 334.0 +- 4.6 : 25000 (+19071,=5383,-546), 87.0 %
23 ba01f4b 2021-06-19 [nn-75980ca503c6.nnue](https://tests.stockfishchess.org/api/nn/nn-75980ca503c6.nnue) 335.6 +- 4.6 : 25000 (+19108,=5359,-533), 87.2 %
24 9b82414 2021-06-21 [nn-190f102a22c3.nnue](https://tests.stockfishchess.org/api/nn/nn-190f102a22c3.nnue) 343.2 +- 4.6 : 25000 (+19328,=5156,-516), 87.6 %
25 b94a651 2021-06-28 [nn-956480d8378f.nnue](https://tests.stockfishchess.org/api/nn/nn-956480d8378f.nnue) 342.9 +- 4.6 : 25000 (+19293,=5218,-489), 87.6 %
26 49283d3 2021-06-28 [nn-3475407dc199.nnue](https://tests.stockfishchess.org/api/nn/nn-3475407dc199.nnue) 348.1 +- 4.7 : 25000 (+19405,=5151,-444), 87.9 %
27 8fc297c 2021-07-03 [nn-9e3c6298299a.nnue](https://tests.stockfishchess.org/api/nn/nn-9e3c6298299a.nnue) 347.2 +- 4.7 : 25000 (+19377,=5172,-451), 87.9 %
28 b939c80 2021-07-24 [nn-76a8a7ffb820.nnue](https://tests.stockfishchess.org/api/nn/nn-76a8a7ffb820.nnue) 358.1 +- 4.9 : 25000 (+19658,=4940,-402), 88.5 %

**Comment by vondele** [comment]
The above Elo gain graphically

![Elo](https://user-images.githubusercontent.com/4202567/126906368-0e303f91-20ed-4445-91a6-f976f13441b8.png)

**Comment by Sopel97** [comment]
Could you publish the script you used to get these plots? I would like to run this for some other configurations and a longer time span

**Comment by vondele** [comment]
I don't think I still have the scripts, except for the one finding the relevant commits:
```bash
#!/bin/bash

#
# find all nets for all commits
#

#
# all commits since classical
#

git checkout master
revs=`git rev-list 84f3e8679^..HEAD | awk '{c++; if (c%1==0) print $0}' | tac`
netold=""
rm -f stockfish.all.net

for rev in $revs
do
  epoch=`git show --pretty=fuller --date=short $rev | grep 'CommitDate' | awk '{print $NF}'`
  git checkout $rev >& /dev/null
  net=`grep EvalFileDefaultName evaluate.h | grep define | sed 's/.*\(nn-[a-z0-9]\{12\}.nnue\).*/\1/'`
  if [ "$net" == "" ]; then
     net=`grep EvalFile ucioption.cpp | grep Option | sed 's/.*\(nn-[a-z0-9]\{12\}.nnue\).*/\1/'`
  fi
  if [ "$net" != "$netold" ]; then
     netold=$net
     make clean && make -j ARCH=x86-64-avx2 profile-build
     mv stockfish stockfish.net.$rev
     echo $rev $epoch $net >> stockfish.all.net
  fi
done
```
after that a simple script to run over all revisions listed in the file `stockfish.all.net` to compute what you need. The data is just imported in a plotting tool (These graphs are using grace).
